### PR TITLE
DEV-1625: {variable} substitution in Mode-A raw-SQL surfaces

### PR DESCRIPTION
**Pull request by ZmeiGorynych**
## What & why

Extends SLayer's `{variable}` substitution to the four **raw-SQL (Mode A) surfaces** of a query's **direct source model** — `SlayerModel.sql`, `SlayerModel.filters`, `Column.sql`, `Column.filter`. This is the missing primitive for parameterizing hand-written model SQL, and the thing needed to represent a Cube `FILTER_PARAMS` cube as an `sql`-mode model ([DEV-1608](https://linear.app/motley-ai/issue/DEV-1608)).

Closes [DEV-1625](https://linear.app/motley-ai/issue/DEV-1625).

## Approach

- **`substitute_variables` gains a required keyword-only `escape: Literal["sql","python"]`.** Mode-A surfaces are sqlglot-parsed → `escape="sql"` doubles `'`→`''`. Mode-B query filters are Python-AST-parsed → `escape="python"` backslash-escapes `\`/`'`/`"` (in that order) so `ast.parse` recovers the value; SQL quote-doubling would be read as adjacent-literal concatenation (`'O''Brien'` → `OBrien`) and silently corrupt it. `bool` is accepted (renders `True`/`False`); non-finite floats (`nan`/`inf`) and unknown `escape` modes raise.
- **`_substitute_model_sql_surfaces`** renders the direct source model in `_execute_pipeline` between `_resolve_query_model` and `_enrich`, guarded by `model.source_model_origin is None` (rendered virtual models skipped). Because the same substituted `model` object flows into `_build_rerooted_enriched`, `model.filters` substitution stays consistent across the main and cross-model re-rooted CTEs. `model_copy`-based (never mutates the cached model); an empty-vars guard leaves variable-free models untouched so raw brace literals (`'{1,2,3}'`) survive.
- **`get_column_types`** gains a defaults-only pass (`model.query_variables` only, no dummy fill) inside the probe's try-boundary, so a fully-defaulted parameterized model resolves real catalog types instead of degrading to `{}`; undefaulted/partially-defaulted models still degrade gracefully.
- Effective vars = `{**(model.query_variables or {}), **(query.variables or {})}`; existing precedence (runtime kwarg > stage > outer > model defaults) is reused. Missing variable → raise. `inspect_model` shows the literal `{var}` template.

## Scope / deferred

Substitution applies to the **direct source model** only. Nested `source_queries` stages, query-backed direct sources, join targets, and cross-model targets are the deferred follow-up ([DEV-1678](https://linear.app/motley-ai/issue/DEV-1678)); a `{var}` in those lineages stays literal and errors on the stray placeholder.

## Notable findings while implementing

- The cross-model re-rooted CTE **drops** a source filter unreachable from the target (pre-existing `drop_unreachable_filters` semantics) — pinned by two tests: an unreachable-target boundary case and a bidirectional-join reachable case.
- **Backslash-in-value does not round-trip through SQLite** even for literal, no-variable filters (the driver doesn't unescape backslashes; the Mode-B parser reads `\b` as backspace) — a pre-existing dialect quirk, not introduced here. Backslash coverage lives at the unit level (escaping contract + AST round-trip); one `xfail(strict=True)` documents the SQLite limitation and auto-promotes if ever fixed.

## Tests

Full non-integration suite green (**6372 passed, 3 xfailed**), lint clean. End-to-end against SQLite for each surface, variable precedence, missing-variable, per-mode escaping (quote/double-quote/backslash matrices + AST round-trip), injection/escaping, and regression pins that Mode-B formula surfaces stay un-substituted.

## Docs

`CLAUDE.md`, `docs/concepts/{models,queries,references}.md`, and the `slayer-models` / `slayer-query` skills.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added variable substitution for model SQL, filters, and column expressions.
  * Supports merged model and query variables with clear precedence rules.
  * Added SQL- and Python-aware escaping for substituted values.
  * Added validation for missing variables, unsupported values, and non-finite numbers.

* **Documentation**
  * Added a variable substitution tutorial, examples, and executable notebook.
  * Documented supported surfaces, escaping, scope, errors, and literal braces.

### CODE REVIEW

File: slayer/core/query.py [hunk 1]
```diff
@@ -38,21 +39,56 @@ def _validate_query_filter_string(formula: str) -> None:
...
+    - ``"python"`` — Mode-B query filters are parsed by SLayer's Python-AST
+      formula parser, where SQL quote-doubling would be read as adjacent-literal
+      concatenation (``'O''Brien'`` → ``'OBrien'``). So backslash is doubled
+      FIRST, then both quote styles are backslash-escaped, matching Python
+      string-literal rules so ``ast.parse`` recovers the original value.
+    """
+    if escape == "sql":
+        return value.replace("'", "''")
+    # python: order matters — double the backslash before escaping quotes.
+    return value.replace("\\", "\\\\").replace("'", "\\'").replace('"', '\\"')
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/core/query.py:61]: Acknowledged. DEV-1625 deliberately scoped injection-safety to "trusted (caller-supplied) input + documented caveat" — `substitute_variables` is dialect-agnostic and doubles single quotes, which is correct for standard-SQL backends but, as you note, insufficient on backslash-escaping dialects like MySQL. I've kept the current behavior for this PR and made the MySQL/trusted-input limitation explicit in `docs/concepts/models.md`. Full dialect-aware escaping (or bound parameters) is tracked as a scoped follow-up in DEV-1727.

File: slayer/engine/query_engine.py [hunk 2]
```diff
@@ -114,6 +115,44 @@ def _merge_query_variables(
...
+
+    No-op copy when ``variables`` is empty — so a model that uses no variables
+    is never touched and raw brace literals (e.g. Postgres arrays ``'{1,2,3}'``)
+    survive verbatim. Uses the hardened :func:`substitute_variables`
+    (raise-on-missing, string single-quote escaping). Never mutates the input
+    model — it may be a shared cached object. Only these four surfaces change;
+    Mode-B surfaces (``ModelMeasure.formula`` etc.) are copied through as-is.
+    """
+    if not variables:
+        return model
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply slayer/engine/query_engine.py:278]: Good catch on the contract gap. This is intentional but was under-documented: the empty-map fast path exists so a fully variable-free model keeps raw brace literals (e.g. Postgres arrays `'{1,2,3}'`) intact. The precise contract is "raise-on-missing once any variable is in play; a fully variable-free execution treats braces as literals." I've made that explicit in `docs/concepts/models.md`, `references.md`, and the slayer-models skill, and added a test pin (`test_placeholder_shaped_literal_in_variable_free_model_not_raised`). The moment any variable is supplied, an unresolved `{var}` still raises (covered by TestMissingVariable). Keeping the behavior avoids a backward-incompatible change to existing brace-literal models.

File: tests/test_mode_a_variable_substitution.py [hunk 3]
```diff
@@ -0,0 +1,1342 @@
...
+            q = SlayerQuery(
+                source_model="orders",
+                measures=[{"formula": "customers.scaled:sum"}],
+                variables={"mult": 2},
+            )
+            # The failure must be about the stray, unsubstituted {mult}
+            # (emitted as a bare `mult` column ref), not an unrelated setup
+            # error — pins that the join-target var truly was not substituted.
+            with pytest.raises(Exception, match="mult"):
+                await engine.execute(q)
```
[the remark this answers is not in the document]
Comment (ZmeiGorynych) [reply tests/test_mode_a_variable_substitution.py]: Tightened, thanks — but not to `ValueError`. I verified the raised exception is a `sqlalchemy.exc.OperationalError` ("no such column: customers.mult"): precisely because the join-target `{mult}` is NOT substituted (deferred to DEV-1678), it surfaces as a bad column ref at the DB, not a substitution `ValueError`. So I kept the broad type (the DB-error class is an implementation detail) and tightened the match to a word boundary `r"\bmult\b"`, which rejects unrelated "multiple"/"multi-stage" while still matching `customers.mult`.
